### PR TITLE
Handle time based conditional requests `fdsnws-station`

### PR DESCRIPTION
**Pull request by damb**
**Features and Changes**:

- Support time based conditional requests (`fdsnws-station` resource) based on [RFC 7232](https://tools.ietf.org/html/rfc7232#section-3.3).

Let me know if you're interested in this feature. If yes, I'm going to cover the changes with additional test cases.

### CODE REVIEW

**Comment by gempa-stephan** [comment]
Hi Daniel,

thanks for your pull request. This is a great feature. Please go ahead and implement the test cases. Also I like to request a configuration parameter which toggles this feature. With so many installations out there I think it is a good idea to disable the switch by default to not break any workflow.

Regards,
Stephan

**Comment by damb** [comment]
Hi @gempa-stephan,

> Also I like to request a configuration parameter which toggles this feature.

How do you want me to name this configuration parameter? Would something like `handleConditionalRequests` be ok for you?

Daniel

**Comment by gempa-stephan** [comment]
Good naming is always the hard part. What about `handleConditionalHeaders`?

**Comment by damb** [comment]
> Good naming is always the hard part. What about `handleConditionalHeaders`?

This is fine for me. I suggested `handleConditionalRequests` due to https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests.

**Comment by damb** [comment]
@gempa-stephan, I implemented a couple of tests. If you can think about some additional useful tests, let me know.

It would be good if you have a look at the code. Let me know if there are things to be changed.

During test implementation I noted a few things / I had some questions (kind of OT):
- The `fdsnws` module loads the module configuration provided under e.g. `~/seiscomp/etc/fdsnws.cfg`. This is not good since tests should be self contained. Would it be an option to use the `--config-file` with a path to an empty file in order to force the module to load with a well-defined configuration?
- Is there a concrete reason, why the `fdsnws` module tests are implemented not relying on a framework such as e.g. [TwistedTrial](https://twistedmatrix.com/trac/wiki/TwistedTrial)? Such frameworks are particular handy when it comes to implementing parameterized tests etc.

Thanks for sharing your thoughts.

**Comment by gempa-stephan** [comment]
> > Good naming is always the hard part. What about `handleConditionalHeaders`?
> 
> This is fine for me. I suggested `handleConditionalRequests` due to https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests.

In this case just go ahead with your suggestion.

**Comment by damb** [comment]
> In this case just go ahead with your suggestion.

Ok. So I'm going to rename the configuration parameter.

Besides, I noted that I need to validate the `Last-Modified` header field. I'm going to fix this.

**Comment by damb** [comment]
Ok. There we go. Ready for review.

File: apps/fdsnws/test/fdsnwstest.py [hunk 1]
```diff
@@ -22,25 +22,24 @@
...
 
 ###############################################################################
 class FDSNWSTest(object):
 
-    #--------------------------------------------------------------------------
+    # --------------------------------------------------------------------------
     def __init__(self, port=8080):
         self.port = port
-        self.url = 'http://localhost:{}/fdsnws'.format(self.port)
+        self.url = "http://localhost:{}/fdsnws".format(self.port)
```
Comment (gempa-stephan) [inline apps/fdsnws/test/fdsnwstest.py:37]: I would rather keep the single quotes here as they represent an unnecessary change. The same applies for all following quote changes.

File: apps/fdsnws/test/fdsnwstest.py [hunk 2]
```diff
@@ -205,122 +200,186 @@ def diff(expected, got, ignoreRanges):
...
+        retCode=200,
+        testID=None,
+        ignoreRanges=None,
+        auth=False,
+        diffContent=True,
+        silent=False,
+        postData=None,
+        reqHeaders=None,
+        respHeaders=None,
+    ):
```
Comment (gempa-stephan) [inline apps/fdsnws/test/fdsnwstest.py:294]: Please don't break each parameter into a new line here. Again an unnecessary change.

File: apps/fdsnws/test/fdsnwstest.py (hunk 1 above)
Comment (damb) [reply apps/fdsnws/test/fdsnwstest.py:37]: @gempa-stephan, reformatting wasn't on purpose, sorry. I'm going to fix this.

Comment (damb) [reply apps/fdsnws/test/fdsnwstest.py]: Ok. The formatting issue should be fixed now. I had auto-formatting enabled ...

**Comment by gempa-stephan** [comment]
I just did a final test on your PR. Unfortunately the new tests fail under UB20/Python3:

<pre>
3/6 Test #3: station ..........................   Passed    6.72 sec
test 4
    Start 4: station-conditional

4: Test command: /usr/bin/python3 "/path/src/base/main/apps/fdsnws/test/test-station-conditional.py"
4: Environment variables: 
4:  PYTHONDONTWRITEBYTECODE=1
4:  LD_LIBRARY_PATH=/path/build/lib
4:  PYTHONPATH=/path/src/base/main/apps/fdsnws/test:/path/build/lib/python
4:  SEISCOMP_ROOT=/path/src/base/main/apps/fdsnws/test/data
4:  SEISCOMP_LOCAL_CONFIG=/path/build/src/base/main/apps/fdsnws/test/.seiscomp
4: Test timeout computed to be: 10000000
4: Traceback (most recent call last):
4:   File "/path/src/base/main/apps/fdsnws/test/fdsnwstest.py", line 49, in __call__
4:     self.test()
4:   File "/path/src/base/main/apps/fdsnws/test/test-station-conditional.py", line 136, in test
4:     self.testHTTP(qs, ct, reqHeaders=reqHeaders,
4:   File "/path/src/base/main/apps/fdsnws/test/fdsnwstest.py", line 243, in testHTTP
4:     self.testHTTPOneShot(url, contentType, data, dataFile, retCode,
4:   File "/path/src/base/main/apps/fdsnws/test/fdsnwstest.py", line 279, in testHTTPOneShot
4:     raise ValueError(
4: ValueError: Missing response header field: 'last-modified'
4: starting FDSNWS service: /usr/bin/python3 /path/src/base/main/apps/fdsnws/test/data/../../fdsnws.py --debug --plugins=dbsqlite3,fdsnxml --database=sqlite3:///path/src/base/main/apps/fdsnws/test/data/seiscomp3.sqlite3 --serveAvailability=true --dataAvailability.enable=true --agencyID=Test --record-url=sdsarchive:///path/src/base/main/apps/fdsnws/test/data/sds --htpasswd=/path/src/base/main/apps/fdsnws/test/data/fdsnws.htpasswd --stationFilter=/path/src/base/main/apps/fdsnws/test/data/stationFilter.cfg --handleConditionalRequests=true
4: waiting for port 8080 to become ready ... OK
4: Testing station service (conditional requests)
4: ('http://localhost:8080/fdsnws/station/1/query?level=station&format=text&network=AM&station=R187C&location=00&channel=EHZ', None, 200, {'If-Modified-Since': b'Wed, 31 Dec 1969 23:00:00 GMT'}, {'last-modified': <function TestStationConditionalRequests.test.<locals>.validateDateString at 0x7fcc45ef0310>}, '/path/src/base/main/apps/fdsnws/test/data/results/station-conditional-1.txt')
4: #1 http://localhost:8080/fdsnws/station/1/query?level=station&format=text&network=AM&station=R187C&location=00&channel=EHZ: stopping FDSNWS service (PID: 679796): . OK
4/6 Test #4: station-conditional ..............***Failed    1.50 sec
test 5
</pre>

Do you have time to investigate?

**Comment by damb** [comment]
Interesting.

> Do you have time to investigate?

Unfortunately, this week not anymore. Earliest next week. Let me know, if I should have a closer look into the issue.

BTW, I'm on
 - Fedora32
 - Python 3.8.6
 - Twisted==20.3.0
 - requests==2.24.0

and tests run successfully.

**Review by gempa-stephan — COMMENTED** [review]
The reason for the failure of the test suite on my system is the system time. I use CEST where your time is probably set to UTC. Please find the suggested fix together with some small syntax remarks inline.

We are about to tag a new release this week. With the suggested changes your PR might find it's way into the release. Otherwise it will be included in the next one.

File: apps/fdsnws/test/test-station-conditional.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                            respHeaders=None, **kwargs):
+            return [(qs, contentType, retCode,
+                     {'If-Modified-Since': datetimeToString(ifModifiedSince)},
+                     respHeaders, dataFile)
+                    for qs in createQueryStrings(respFormat, levels,
+                                                 **kwargs)]
+
+        print('Testing station service (conditional requests)')
+
+        epoch = time.mktime(time.gmtime(0))
```
[the remark this answers is not in the document]
Comment (damb) [reply apps/fdsnws/test/test-station-conditional.py]: @gempa-stephan, you were totally right - my time is set to UTC.

[the remark this answers is not in the document]
Comment (damb) [reply apps/fdsnws/test/test-station-conditional.py]: Though, the actual issue most probably is located somewhere else. That is, `Wed, 31 Dec 1969 23:00:00 GMT` is according to [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.1) a valid HTTP-date. Thus, the implementation must handle it correctly, and therefore the `Last-Modified` header field must be sent along with the HTTP status code **200** response.

- Could you check the value of the `self._timeInventoryLoaded` member variable in `class FDSNStation`?
- Also we need to make sure that the `_isConditionalRequest` method returns `True` in case of HTTP-dates < *epoch*.

**Comment by gempa-stephan** [comment]
Excellent, we are ready to merge. Please squash all commits into a single one.

**Comment by damb** [comment]
> Excellent, we are ready to merge. Please squash all commits into a single one.

@gempa-stephan, have you seen https://github.com/SeisComP/main/pull/17#discussion_r635806526?

**Comment by gempa-stephan** [comment]
Thank you for your contribution!
